graphene-writer/src/main/kotlin/com/graphene/writer/store/key/elasticsearch/handler/AbstractElasticsearchKeyStoreHandler.kt

```diff
@@ -30,124 +29,188 @@ abstract class AbstractElasticsearchKeyStoreHandler(
+    this.templateIndexPattern = property.templateIndexPattern
+    this.batchSize = property.bulk.actions
+    this.preloadBatchSize = property.bulk.preloadActions
+    this.flushInterval = property.bulk.interval
+    this.keyCache = SimpleLocalKeyCache(property.cacheExpireIntervalInSeconds)
+    this.cacheExpireIntervalInSeconds = property.cacheExpireIntervalInSeconds
+    this.targetProcessTime = property.targetProcessTime
+
+    this.elasticsearchClient = elasticsearchClient(keyStoreHandlerProperty, elasticsearchClientFactory, property)
+    this.elasticsearchClient.createTemplateIfNotExists(templateIndexPattern, templateName(), templateSource())
+    this.elasticsearchClient.createIndexIfNotExists(setOf(elasticsearchClient.getIndexWithCurrentDate(index, tenant)))
+
+    this.keyStoreScheduler = Executors.newSingleThreadScheduledExecutor(NamedThreadFactory(this::class.simpleName!!))
+    this.keyStoreScheduler.scheduleWithFixedDelay(this, 3_000L, this.flushInterval, TimeUnit.MILLISECONDS)
+
+    val requestOptionsBuilder = RequestOptions.DEFAULT.toBuilder()
+    requestOptionsBuilder.setHttpAsyncResponseConsumerFactory(HttpAsyncResponseConsumerFactory
+      .HeapBufferedResponseConsumerFactory(1000 * 1024 * 1024))
+    this.requestOptions = requestOptionsBuilder.build()
   }
 
-  private fun elasticsearchClient(keyStoreHandlerProperty: KeyStoreHandlerProperty, elasticsearchClientFactory: ElasticsearchClientFactory, property: ElasticsearchKeyStoreHandlerProperty): ElasticsearchClient {
-    return elasticsearchClientFactory.createElasticsearchClient(keyStoreHandlerProperty.rotation, property.cluster, property.port, property.userName, property.userPassword, property.protocol)
+  private fun elasticsearchClient(
+    keyStoreHandlerProperty: KeyStoreHandlerProperty,
+    elasticsearchClientFactory: ElasticsearchClientFactory,
+    property: ElasticsearchKeyStoreHandlerProperty
+  ): ElasticsearchClient {
+    return elasticsearchClientFactory.createElasticsearchClient(
+      keyStoreHandlerProperty.rotation,
+      property.cluster,
+      property.port,
+      property.userName,
+      property.userPassword,
+      property.protocol
+    )
   }
 
   override fun handle(grapheneMetric: GrapheneMetric) {
-    metrics.offer(grapheneMetric)
+    val index = elasticsearchClient.getIndexWithDate(this.index, this.tenant, grapheneMetric.timestampMillis())
+    if (isProcessable(grapheneMetric) && keyCache.putIfAbsent("${index}_${grapheneMetric.id}")) {
+      metrics.offer(grapheneMetric)
+    }
+    handlePreload(index, grapheneMetric)
+  }
+
+  private fun handlePreload(originalIndex: String, grapheneMetric: GrapheneMetric) {
+    val preloadIndex = elasticsearchClient.getIndexWithDate(
+      index,
+      tenant,
+      grapheneMetric.timestampMillis() + (this.cacheExpireIntervalInSeconds * 1_000L)
+    )
+    if (originalIndex != preloadIndex && isProcessable(grapheneMetric) && keyCache.putIfAbsent("${preloadIndex}_${grapheneMetric.key}")) {
```

Comment:
How about caching by id like above?

```suggestion
    if (originalIndex != preloadIndex && isProcessable(grapheneMetric) && keyCache.putIfAbsent("${preloadIndex}_${grapheneMetric.id}")) {
```

Reply:
Oh, that's important thing it was my mistake. Thanks for the advice.